### PR TITLE
Ensure all ref local and field value access are dereferenced

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/62122. Today, we are not dereferencing locals or ref fields if the result of the access is unused (for example, with an explicit `_ =` assignment). This means that side effects, such as a `NullReferenceException` are not observed as they should be.

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
Should this be documented as a breaking change?

**Comment by 333fred** [comment]
> Should this be documented as a breaking change?

This is still in draft for a reason :)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @cston for review.

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md [hunk 1]
```diff
@@ -1,5 +1,16 @@
...
+## Unused results from ref local are dereferences.
+
+***Introduced in Visual Studio 2022 version 17.4***
+
+When a `ref` local variable is referenced by value, but the result is not used (such as being assigned to a discard), the result was previously ignored. The compiler will now dereference that local, ensuring that any side effects are observed.
+
+```csharp
+ref int local = Unsafe.NullRef<int>();
+_ = local; // Will now produce a `NullReferenceException`
+```
```
Comment (jaredpar) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 7.md:12]: User asked for Unsafe. User got Unsafe. 😄

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefReadonlyReturnTests.cs [hunk 2]
```diff
@@ -392,14 +394,16 @@ static void M()
...
-  IL_0005:  pop
-  IL_0006:  ret
-}");
+  IL_0000:  call       "ref int C.Helper()"
+  IL_0005:  ldind.i4
+  IL_0006:  pop
+  IL_0007:  ret
+}
+""");
         }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefReadonlyReturnTests.cs:407]: > } [](http://example.com/codeflow?start=7&length=2)

Consider adding a test with two `ref` locals that are unused, and execute the test.
```
ref readonly int x = ref Helper();
ref readonly int y = ref Helper();
_ = x + y;
```

**Review by cston — APPROVED** [review]

**Review by cston — APPROVED** [review]

**Pull request merged**